packages/transactions/src/new-signatures.ts

```diff
@@ -0,0 +1,98 @@
+import { Address, getAddressFromPublicKey } from '@solana/addresses';
+import { Decoder } from '@solana/codecs-core';
+import { getBase58Decoder } from '@solana/codecs-strings';
+import {
+    SOLANA_ERROR__TRANSACTION__FEE_PAYER_SIGNATURE_MISSING,
+    SOLANA_ERROR__TRANSACTION__SIGNATURES_MISSING,
+    SolanaError,
+} from '@solana/errors';
+import { Signature, SignatureBytes, signBytes } from '@solana/keys';
+
+import { NewTransaction } from './new-compile-transaction';
+
+export interface FullySignedTransaction extends NewTransaction {
+    readonly __brand: unique symbol;
+}
+
+let base58Decoder: Decoder<string> | undefined;
+
+export function newGetSignatureFromTransaction(transaction: NewTransaction): Signature {
+    if (!base58Decoder) base58Decoder = getBase58Decoder();
+
+    // We have ordered signatures from the compiled message accounts
+    // first signature is the fee payer
+    const signatureBytes = Object.values(transaction.signatures)[0];
+    if (!signatureBytes) {
+        throw new SolanaError(SOLANA_ERROR__TRANSACTION__FEE_PAYER_SIGNATURE_MISSING);
+    }
+    const transactionSignature = base58Decoder.decode(signatureBytes);
+    return transactionSignature as Signature;
+}
+
+function uint8ArraysEqual(arr1: Uint8Array, arr2: Uint8Array) {
+    return arr1.length === arr2.length && arr1.every((value, index) => value === arr2[index]);
+}
+
+export async function newPartiallySignTransaction(
+    keyPairs: CryptoKeyPair[],
+    transaction: NewTransaction,
+): Promise<NewTransaction> {
+    let newSignatures: Record<Address, SignatureBytes> | undefined;
+
+    await Promise.all(
+        keyPairs.map(async keyPair => {
+            const address = await getAddressFromPublicKey(keyPair.publicKey);
+            const newSignature = await signBytes(keyPair.privateKey, transaction.messageBytes);
+            const existingSignature = transaction.signatures[address];
+
+            // what to do if existingSignature is undefined, ie not a valid signer?
```

Comment:
I think you should drop the signer up there at the point where you've derived the public key. If `address` is not a key of `transaction.signatures` then either silently drop it or throw something like ‘hey this keypair isn't an eligible signer.’

Reply:
Of note: we throw in the existing library.

https://github.com/solana-labs/solana-web3.js/blob/a98958b63f203613b84cd6b905765debab46ae51/packages/library-legacy/src/transaction/legacy.ts#L753-L755

Reply:
I was undecided whether to throw but I've added an error for this. It includes all the unexpected signers and also all the expected ones in the error context. I've hit that error a lot and that's what I tend to end up using to figure it out!